exercises/affine-cipher/canonical-data.json

```diff
@@ -0,0 +1,191 @@
+      "comments": [ "Test encoding from English to ciphertext with keys" ],
+      "cases": [
+        {
+          "description": "encode yes",
+          "property": "encode",
+          "input": {
+            "phrase": "yes",
+            "key": {
+              "a": 5,
+              "b": 7
+            }
+          },
+          "expected": "xbt"
+        },
+        {
+          "description": "encode no",
+          "property": "encode",
+          "input": {
+            "phrase": "no",
+            "key": {
+              "a": 15,
+              "b": 18
+            }
+          },
+          "expected": "fu"
+        },
+        {
+          "description": "encode OMG",
+          "property": "encode",
+          "input": {
+            "phrase": "OMG",
+            "key": {
+              "a": 21,
+              "b": 3
+            }
+          },
+          "expected": "lvz"
+        },
+        {
+          "description": "encode O M G",
+          "property": "encode",
+          "input": {
+            "phrase": "O M G",
+            "key": {
+              "a": 25,
+              "b": 47
+            }
+          },
+          "expected": "hjp"
+        },
+        {
+          "description": "encode mindblowingly",
+          "property": "encode",
+          "input": {
+            "phrase": "mindblowingly",
+            "key": {
+              "a": 11,
+              "b": 15
+            }
+          },
+          "expected": "rzcwa gnxzc dgt"
+        },
+        {
+          "description": "encode numbers",
+          "property": "encode",
+          "input": {
+            "phrase": "Testing,1 2 3, testing.",
+            "key": {
+              "a": 3,
+              "b": 4
+            }
+          },
+          "expected": "jqgjc rw123 jqgjc rw"
+        },
+        {
+          "description": "encode deep thought",
+          "property": "encode",
+          "input": {
+            "phrase": "Truth is fiction.",
+            "key": {
+              "a": 5,
+              "b": 17
+            }
+          },
+          "expected": "iynia fdqfb ifje"
+        },
+        {
+          "description": "encode all the letters",
+          "property": "encode",
+          "input": {
+            "phrase": "The quick brown fox jumps over the lazy dog.",
+            "key": {
+              "a": 17,
+              "b": 33
+            }
+          },
+          "expected": "swxtj npvyk lruol iejdc blaxk swxmh qzglf"
+        },
+        {
+          "description": "encode with a not coprime to m",
+          "property": "encode",
+          "input": {
+            "phrase": "This is a test.",
+            "key": {
+              "a": 6,
+              "b": 17
+            }
+         },
+          "expected": "Error: a and m must be coprime."
+        }
+      ]
+    },
+    {
+      "description": "decode",
+      "comments": [ "Test decoding from ciphertext to English with keys" ],
+      "cases": [ 
```

Comment:
There should be a test decoding encoded text longer than five characters which does not have spaces. 

I'd expect such a test to require that the decoder successfully decode the input even without the spaces, but that's arguable; one could reasonably say that it should throw an error instead. However, right now, we're not defining any behavior in that case by our tests.

Reply:
This type of test is one that I think each track team should add if they want.
- I have a number of tests that I did not include here because they test to see if input can be piped into the command, and if the program can accept multiline files as an argument.
  - those tests are bash specific, so while I will submit to the bash track they would be inappropriate here.
- Additionally, returning an error if the spaces aren't right in the input would be a nightmare in Bash.
  - It might have a quick (although not very readable) solution with `sed` or `awk`, but the guidelines for the bash track stipulate that we shouldn't use those external programs.
#### If some languages want to add that I think they should, but for some languages it will make the program unnecessarily messy and hard to read.
- most solutions are going to strip out all non-alphanumeric characters for encoding and decoding
- on my host machine I made another script that uses my bash Affine.sh to test the possible keys on any given string or file and spit out the decryption with the keys tagged.
  - That way if I forget the key to something I encrypted I can find it again.
  - Also, if someone has opened an encrypted file and inadvertently added or deleted a character, it is much more pragmatic for the program to decode the text with the error instead of throwing an error and forcing the user to look through encoded text to fix the issue before decoding.

Reply:
I think a test to show that encoded text without spaces can be decoded successfully makes more sense than throwing an error on non-spaced or incorrectly spaced input.  
Most sensible solutions should already do this though.

Reply:
Agree that it makes more sense to expect text without spaces to decode correctly than to expect an error in that case. Agree that most sensible solutions should already do that.

However, behavior which is untested is undefined, and undefined behavior is the devil. I can't speak for why the Atbash cipher doesn't include a test case for that, but it doesn't matter: we're working on this test suite, not that one. We can do better.

Reply:
Okay, thank you for the response.  I didn't see this earlier.